[java] Revamp GuardLogStatementRule to allow var, field and array accesses

jsotuyod opened the pull request:
## Describe the PR

<!-- A clear and concise description of the bug the PR fixes or the feature the PR introduces. -->

## Related issues

- Fixes #3602 
- Fixes #4731
- Fixes #5151
- Fixes #5152
- Fixes #5153

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: Compared to master:<br />
This...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 13 violations, 2 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/10444229280?pr=5169">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oowekyala commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/GuardLogStatementRule.java:
```diff
@@ -237,7 +240,14 @@ private boolean isUnguardedAccessOk(ASTMethodCall call, int messageArgIndex) {
+        return it instanceof ASTStringLiteral || it instanceof ASTLambdaExpression
+                || it instanceof ASTVariableAccess || it instanceof ASTMethodReference
+                || it instanceof ASTFieldAccess || it instanceof ASTThisExpression
+                || (it instanceof ASTArrayAccess && isDirectAccess(((ASTArrayAccess) it).getQualifier()));
```
Checking that the qualifier is a direct access should also be done for method references and field accesses (all QualifiableExpressions) : `foo().field`, `foo()::ref`

oowekyala replied:
I think it's illegal to use anything else than a class type as the LHS of a this expression, so I'm not sure it's worth checking for that

jbisotti commented:
It would be great if the documentation could also be expanded to explain the nuances here. Maybe include examples like in my original Issues that should pass the check and then slight variations of them that would not. Thanks.

jsotuyod commented:
@jbisotti I changed the rule description slightly and added some extra examples, feel free to comment.

The bottom line is, we should flag any "expensive operations", with expensive being one of:
- string manipulation / concatenation
- method invocations
- constructor invocations

The last 2 are specially dangerous as the methods may also have side-effects, so you are using a log call to actually perform business logic.

Accessing fields / variables / array elements / method references directly is ok. Doing that from the return of a method invocation / constructor call is not.

adangel reviewed (approved):
Thanks!
